From 72b4beac5e62e5855e3245ccba5a133c10dff1d4 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Mon, 24 Jul 2023 20:16:17 +0800
Subject: [PATCH 24/53] [Huawei] mm/sharepool: Simplify mg_sp_free()

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Delete useless wrap function and the state machine used in mg_sp_free().

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 64 +++++++++++++------------------------------------
 1 file changed, 16 insertions(+), 48 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 8f2b12b0c86b..f5185b586a78 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1911,29 +1911,9 @@ static void sp_fallocate(struct sp_area *spa)
 		WARN(1, "sp fallocate failed %d\n", ret);
 }
 
-static void sp_free_unmap_fallocate(struct sp_area *spa)
-{
-	down_read(&spa->spg->rw_lock);
-	__sp_free(spa->spg, spa->va_start, spa_size(spa), NULL);
-	sp_fallocate(spa);
-	up_read(&spa->spg->rw_lock);
-}
-
-#define FREE_CONT	1
-#define FREE_END	2
-
-struct sp_free_context {
-	unsigned long addr;
-	struct sp_area *spa;
-	int state;
-	int spg_id;
-};
-
-/* when success, __sp_area_drop(spa) should be used */
-static int sp_free_get_spa(struct sp_free_context *fc)
+static int sp_free_inner(unsigned long addr, int spg_id)
 {
 	int ret = 0;
-	unsigned long addr = fc->addr;
 	struct sp_area *spa;
 	struct sp_group *spg;
 
@@ -1944,19 +1924,17 @@ static int sp_free_get_spa(struct sp_free_context *fc)
 		return -EINVAL;
 	}
 
-	spg = sp_group_get_from_mm(current->mm, fc->spg_id);
+	spg = sp_group_get_from_mm(current->mm, spg_id);
 	if (!spg) {
-		pr_err("sp free get group failed %d\n", fc->spg_id);
+		pr_err("sp free get group failed %d\n", spg_id);
 		return -EINVAL;
 	}
 
-	fc->state = FREE_CONT;
-
 	spa = get_sp_area(spg, addr);
-	sp_group_put(spg);
 	if (!spa) {
 		pr_debug("sp free invalid input addr %lx\n", addr);
-		return -EINVAL;
+		ret = -EINVAL;
+		goto out;
 	}
 
 	if (spa->type != SPA_TYPE_ALLOC) {
@@ -1964,15 +1942,14 @@ static int sp_free_get_spa(struct sp_free_context *fc)
 		pr_debug("sp free failed, %lx is not sp alloc addr\n", addr);
 		goto drop_spa;
 	}
-	fc->spa = spa;
 
 	down_write(&spa->spg->rw_lock);
 	if (!spg_valid(spa->spg)) {
-		fc->state = FREE_END;
 		up_write(&spa->spg->rw_lock);
 		goto drop_spa;
 		/* we must return success(0) in this situation */
 	}
+
 	/* the life cycle of spa has a direct relation with sp group */
 	if (unlikely(spa->is_dead)) {
 		up_write(&spa->spg->rw_lock);
@@ -1984,10 +1961,18 @@ static int sp_free_get_spa(struct sp_free_context *fc)
 	spa->is_dead = true;
 	up_write(&spa->spg->rw_lock);
 
-	return 0;
+	down_read(&spa->spg->rw_lock);
+	__sp_free(spa->spg, spa->va_start, spa_size(spa), NULL);
+	sp_fallocate(spa);
+	up_read(&spa->spg->rw_lock);
+
+	if (current->mm != NULL)
+		sp_update_process_stat(current, false, spa);
 
 drop_spa:
 	__sp_area_drop(spa);
+out:
+	sp_group_put(spg);
 	return ret;
 }
 
@@ -2005,12 +1990,6 @@ static int sp_free_get_spa(struct sp_free_context *fc)
  */
 int mg_sp_free(unsigned long addr, int id)
 {
-	int ret = 0;
-	struct sp_free_context fc = {
-		.addr = addr,
-		.spg_id = id,
-	};
-
 	if (!sp_is_enabled())
 		return -EOPNOTSUPP;
 
@@ -2019,18 +1998,7 @@ int mg_sp_free(unsigned long addr, int id)
 	if (current->flags & PF_KTHREAD)
 		return -EINVAL;
 
-	ret = sp_free_get_spa(&fc);
-	if (ret || fc.state == FREE_END)
-		goto out;
-
-	sp_free_unmap_fallocate(fc.spa);
-
-	if (current->mm != NULL)
-		sp_update_process_stat(current, false, fc.spa);
-
-	__sp_area_drop(fc.spa);  /* match get_sp_area in sp_free_get_spa */
-out:
-	return ret;
+	return sp_free_inner(addr, id);
 }
 EXPORT_SYMBOL_GPL(mg_sp_free);
 
-- 
2.17.1

